### DISCUSSION TITLE
Feedback on v1.6.0 Release

### DISCUSSION THREAD

**Discussion by pthom** [pthom/imgui_bundle#282] [Announcements]
The v1.6.0 release includes significant updates like the update to ImGui 1.91.5, the transition to Nanobind for Python bindings, Python 3.13 support, improved node editor features, and several quality-of-life enhancements across the library.

To ensure this update meets your expectations and continues to improve your experience, we’d love to hear your feedback! 

Here are some subject for which your feedback would be appreciated:

- **Python Bindings**: How has the switch to Nanobind impacted your workflow? Have you encountered any issues?
- **Node Editor Improvements**: Do you feel the updated popup placement and theme consistency are helpful?
- **ImmVision Breaking Change**: What do you think about the requirement to explicitly set the color order? Was the migration smooth?

Feel free to share any other thoughts, suggestions, or bug reports as well. Your input is very valuable!

Thanks!

**Comment by bgribble** [comment]
Looks good! Thanks for all your work in getting this release together!

There were a few compatibility issues that were crashers for me on upgrade, not sure if they are changes in imgui or in the binding wrappers:

* `imgui.Key.im_gui_mod_<name>` are missing, now are `imgui.Key.mod_<name>`
* `imgui.Key.count` has been renamed `imgui.Key.named_key_end`
* `imgui.input_double` now rejects the flag `ImGuiInputTextFlags_EnterReturnsTrue`, which makes editing the text of a double a little less convenient (can't as easily control when the value gets "posted")
* There's a weird rendering problem on `imgui_node_editor` when I am abusing it to drag the viewport around with the mouse, I need to check into this a little more (to do this, I have to create a couple of invisible nodes and then use the `zoom_to_selection` API... not entirely surprised that it breaks). Screencast attached


https://github.com/user-attachments/assets/e1b00713-3bf7-403f-b47c-8afbc6375c79

**Comment by pthom** [comment]
@bgribble 

> Looks good! Thanks for all your work in getting this release together!

Thank you!

> imgui.Key.im_gui_mod_<name> are missing, now are imgui.Key.mod_<name>

=> This is a change in the wrapper, I removed an annoying prefix

> imgui.Key.count has been renamed imgui.Key.named_key_end
> imgui.input_double now rejects the flag ImGuiInputTextFlags_EnterReturnsTrue, which makes editing the text of a double a little less convenient (can't as easily control when the value gets "posted")

=> those are changes in the ImGui API.

> There's a weird rendering problem on imgui_node_editor when I am abusing it to drag the viewport around with the mouse, I need to check into this a little more (to do this, I have to create a couple of invisible nodes and then use the zoom_to_selection API... not entirely surprised that it breaks). Screencast attached

I'm not sure this is related to this update. Can you please try it on the previous version, and tell me if this is a new issue.

**Reply by bgribble** [reply]
Thanks for clarifying on the `Key` changes, I approve of removing the annoying prefix, just wanted to flag that it was a breaking change for me.

> I'm not sure this is related to this update.

It's definitely a new behavior since the last published version of `imgui_bundle` (I am not tracking github, just whatever is in pypi).

**Reply by pthom** [reply]
Can you extract a minimal reproducer code ?

**Comment by bgribble** [comment]
Another new behavior: calling `imgui.get_clipboard_text()` throws the exception `RuntimeError: std::bad_cast`. 

I am using the SDL2 backend and I notice [this ChangeLog](https://github.com/ocornut/imgui/blob/5b7feebfd8f3e58392d955b4c8ae01f77eae25ed/backends/imgui_impl_sdl2.cpp#L26C1-L30C81) entry that indicates something changed in Imgui regarding this functionality.

**Comment by pthom** [comment]
@bgribble : Thanks for letting me know!

I could not reproduce the bad_cast error. However, I saw a related issue which happens when the clipboard actually does not contain a string, but an image (or something else which is not text): in this case, we had a segfault.

I solved this in [this commit](https://github.com/pthom/imgui/commit/216066119500a2f4b46f0271c7b94ab286376500) in the imgui fork.

You can download wheels with the correction from [here](https://github.com/pthom/imgui_bundle/actions/runs/11989960580) (they should be ready in 1 hour).

Can you test them and tell me if that solves the issue? BTW, on which platform are you running?

**Reply by pthom** [reply]
ping @bgribble

**Reply by bgribble** [reply]
Hi, I installed the wheel from your build artifacts linked above. No change in behavior.

Since my code is triggered possibly outside the imgui render() loop, I also tried adding this little code inside the main render:

```
        log.debug("setting clipboard")
        imgui.set_clipboard_text("hello, world")
        tt = imgui.get_clipboard_text()
        log.debug(f"got clipboard: {tt}")
```

Same result -- `std::bad_cast` error on the `get_clipboard_text()` call

I am on Python 3.12 on 64-bit Intel Linux (debian unstable)

**Reply by pthom** [reply]
Can you provide me with a minimal reproducible example code ?

**Reply by bgribble** [reply]
This is a diff against  `example_python_backend_sdl2.py` from the source tree which demostrates the `bad_cast` error.

The SetHint changes are needed for the demo to run at all on my machine 

```
*** /home/grib/Downloads/example_python_backend_sdl2.py 2024-12-03 08:09:01.178069813 -0500
--- example_python_backend_sdl2.py      2024-12-03 08:08:33.721877833 -0500
***************
*** 76,81 ****
--- 76,85 ----
                      {io.key_super=}""")
              imgui.end()

+         imgui.set_clipboard_text("Hello, world!")
+         tt = imgui.get_clipboard_text()
+         print("clipboard: '{tt}'")
+
          gl.glClearColor(1.0, 1.0, 1.0, 1)
          gl.glClear(gl.GL_COLOR_BUFFER_BIT)

***************
*** 113,118 ****
--- 117,124 ----

      SDL_SetHint(SDL_HINT_MAC_CTRL_CLICK_EMULATE_RIGHT_CLICK, b"1")
      SDL_SetHint(SDL_HINT_VIDEO_HIGHDPI_DISABLED, b"1")
+     SDL_SetHint(SDL_HINT_VIDEODRIVER, b"wayland,x11")
+     SDL_SetHint(SDL_HINT_VIDEO_X11_FORCE_EGL, b"1")

      window = SDL_CreateWindow(
          window_name.encode("utf-8"),
```

**Reply by pthom** [reply]
This should be fixed by https://github.com/pthom/imgui_bundle/commit/6e561c2210d324965335b6c16ae787d67bb41532

**Reply by bgribble** [reply]
Confirmed! Thanks for the quick fix
